### PR TITLE
add mixed data type support for GroupNorm backward on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89485
* __->__ #88663

### Motivation
Amp provides convenience methods for mixed precision. If users use amp to run bfloat16 models, torch.autocast will keep module parameters in acc dtype which will leave gamma and beta in float while input/output will be in bfloat16. The same goes for backward: parameters are in float, and X & dX & dY are in bfloat16.
Mixed data type support for GroupNorm backward is also needed for model training with GroupNorm.

### Testing

Single socket (28cores):
* Contiguous:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 3.08E-05 | 3.50E-05 | 8.06E-05 | 7.69E-05
[10, 128, 50, 50] | 0.000121 | 0.000114 | 0.000358 | 0.000203

* Channels Last (inputs and outputs will be converted to contiguous):

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 4.04E-05 | 4.41E-05 | 0.000226 | 0.000305
[10, 128, 50, 50] | 0.000169 | 0.000166 | 0.001628 | 0.001169

Single core:

* Contiguous:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 2.38E-04 | 2.51E-04 | 5.94E-04 | 4.50E-04
[10, 128, 50, 50] | 0.00171 | 0.001395 | 0.0044455 | 0.00243

* Channels Last (inputs and outputs will be converted to contiguous):

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 2.28E-04 | 3.26E-04 | 0.0016528 | 0.003165
[10, 128, 50, 50] | 0.001788 | 0.001302 | 0.0276621 | 0.019447

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 1]
```diff
@@ -316,16 +417,16 @@ void GroupNormKernelImplChannelsLastInternal(
...
             rstd_val += buffer_ptr[g * D + d + C];
            }
         }
         mean_val *= s;
         rstd_val = std::max(rstd_val * s - mean_val * mean_val, T_ACC(0));
         rstd_val = T_ACC(1) / std::sqrt(rstd_val + eps);
-        mean_data[n * G + g] = T(mean_val);
-        rstd_data[n * G + g] = T(rstd_val);
+        tmp_buffer_data[n * 2 * G + 2 * g] = mean_val;
+        tmp_buffer_data[n * 2 * G + 2 * g + 1] = rstd_val;
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:429]: May I know the reason why we don't store the mean/rstd to mean_data and rstd_data directly but use a temp buffer to do it?

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 2]
```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
...
+  float* acc_buffer = new float[C];
+  at::parallel_for(0, C, K, [=](int64_t start, int64_t end) {
+    std::memset(acc_buffer + start, 0, (end - start) * sizeof(float));
+    for (const auto i : c10::irange(N)) {
+      const float* db_ptr = db + i * C;
+      for (const auto j : c10::irange(start, end)) {
+        acc_buffer[j] += db_ptr[j];
+      }
+    }
+  });
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:821]: ditto: reuse unspecialized version?

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 3]
```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
...
+    const BFloat16* rstd,
+    const float* ds,
+    const float* db,
+    BFloat16* dgamma) {
+  using fVec = vec::Vectorized<float>;
+  using bVec = vec::Vectorized<BFloat16>;
+  const int64_t G = group;
+  const int64_t D = C / G;
+  constexpr int64_t K = bVec::size();
+  float* acc_buffer = new float[C];
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:760]: better to use unique_ptr?

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 4]
```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
...
   });
 }
 
-template <typename T>
+template <>
+void BetaBackward(int64_t N, int64_t C, const float* db, BFloat16* dbeta) {
+  using fVec = vec::Vectorized<float>;
+  using bVec = vec::Vectorized<BFloat16>;
+  constexpr int64_t K = bVec::size();
+  float* acc_buffer = new float[C];
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:812]: ditto

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp:429]: Because if dtype of mean or rstd is bfloat16, it will loss precision.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 5]
```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
...
+    for (int64_t i = 0; i < N * G; ++i) {
+      const float* ds_ptr = ds + i * D;
+      const float* db_ptr = db + i * D;
+      const int64_t g = i % G;
+      for (const auto j : c10::irange(start, end)) {
+        const int64_t c = g * D + j;
+        acc_buffer[c] += (ds_ptr[j] - db_ptr[j] * float(mean[i])) * float(rstd[i]);
+      }
+    }
+  });
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp:774]: It is hard to reuse unspecialized version. If PT is bfloat16, `dgamma[c] += (ds_ptr[j] - db_ptr[j] * T_ACC(mean[i])) * T_ACC(rstd[i])` will get wrong results.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Remove the specialized version for PT = BFloat16 since parameters should not be bfloat16 in training

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 4 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Remove the specialized version for PT = BFloat16 since parameters should not be bfloat16 in training

**Review by mingfeima — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 6]
```diff
@@ -640,10 +798,20 @@ void GroupNormBackwardKernelImpl(
...
     Tensor& dbeta) {
+  const bool mixed_type = is_mixed_type(dY, gamma);
   AT_DISPATCH_FLOATING_TYPES_AND(
       ScalarType::BFloat16, X.scalar_type(), "GroupNormBackwardKernelImpl", [&]() {
-        GroupNormBackwardKernelImplInternal<scalar_t>(
+      // In training, using Amp to enable BFloat16 is recommended.
+      // It will keep module parameters in acc dtype i.e. float
+      // while input/output will be in BFloat16.
+      // Using parameters in BFloat16 will cause high precision loss.
+      if(mixed_type) {
```
Comment (malfet) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:808]: Same feedback as for https://github.com/pytorch/pytorch/pull/81852 : please consider who this could be constrained to just `BFloat16`, perhaps it could be handled outside of the generic switch statement

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 6 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp:815]: Fixed in https://github.com/pytorch/pytorch/pull/89485

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 7]
```diff
@@ -468,64 +565,126 @@ void ComputeInternalGradients(
...
-        ds_vec = ds_vec + vec::Vectorized<T>::loadu(ds_ptr + j) * gamma_vec;
-        db_vec = db_vec + vec::Vectorized<T>::loadu(db_ptr + j) * gamma_vec;
-      }
-      ds_vec.store(ds_arr.data());
-      db_vec.store(db_arr.data());
-      T ds_val = std::accumulate(ds_arr.cbegin(), ds_arr.cend(), T(0));
-      T db_val = std::accumulate(db_arr.cbegin(), db_arr.cend(), T(0));
+      const PT* ds_ptr = ds + i * D;
+      const PT* db_ptr = db + i * D;
+      const PT* gamma_ptr = gamma + g * D;
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp:667]: Thanks for the comments. The CalcDsDb method also checks whether gamma is null.  Will modify this in https://github.com/pytorch/pytorch/pull/100234.
